File: snake/server/app/src/main/java/server/messages/ServerGameOverMessage.java
package server.messages;

import java.time.LocalDateTime;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonSyntaxException;

import server.LocalDateTimeTypeAdapter;

public class ServerGameOverMessage implements Message {

    private String messageCode;
    private boolean winnerExists;
    private String winnerName;
    private LocalDateTime timeStamp;

    public String toJson() {
        Gson gson = new GsonBuilder()
                .registerTypeAdapter(LocalDateTime.class, new LocalDateTimeTypeAdapter())
                .create();
        return gson.toJson(this);
    }

    public ServerGameOverMessage fromJson(String json) throws JsonSyntaxException {

        Gson gson = new GsonBuilder()
                .registerTypeAdapter(LocalDateTime.class, new LocalDateTimeTypeAdapter())
                .create();
        return gson.fromJson(json, ServerGameOverMessage.class);
    }

    public String getMessageCode() {
        return messageCode;
    }

    public void setMessageCode(String messageCode) {
        this.messageCode = messageCode;
    }

    public boolean isWinnerExists() {
        return winnerExists;
    }

    public void setWinnerExists(boolean winnerExists) {
        this.winnerExists = winnerExists;
    }

    public String getWinnerName() {
        return winnerName;
    }

    public void setWinnerName(String winnerName) {
        this.winnerName = winnerName;
    }

    public LocalDateTime getTimeStamp() {
        return timeStamp;
    }

    public void setTimeStamp(LocalDateTime timeStamp) {
        this.timeStamp = timeStamp;
    }

}

File: snake/client/scripts/messages/ServerGameOverMessage.js
class ServerGameOverMessage {
    constructor() {
      this.messageCode = '';
      this.winnerExists = false;
      this.winnerName = '';
      this.timeStamp = null; // Use JavaScript Date object instead of LocalDateTime
    }
  
    toJson() {
      const json = {
        messageCode: this.messageCode,
        winnerExists: this.winnerExists,
        winnerName: this.winnerName,
        timeStamp: this.timeStamp ? TimeFormatter.formatDateTime(this.timeStamp) : null,
      };
      return JSON.stringify(json);
    }
  
    static fromJson(json) {
      const obj = JSON.parse(json);
      const message = new ServerGameOverMessage();
      message.messageCode = obj.messageCode;
      message.winnerExists = obj.winnerExists;
      message.winnerName = obj.winnerName;
      message.timeStamp = obj.timeStamp ? TimeFormatter.parseDateTime(obj.timeStamp) : null;
      return message;
    }
}